src/application_settings/container_base.py

```diff
@@ -0,0 +1,267 @@
+# pylint: disable=consider-alternative-union-syntax
+"""Base classes for containers and sections for configuration and settings."""
+import json
+import sys
+from abc import ABC, abstractmethod
+from dataclasses import asdict, fields, replace
+from enum import Enum
+from pathlib import Path
+from re import sub
+from typing import Any, TypeVar
+
+import tomli_w
+from pathvalidate import is_valid_filepath
+from pydantic.dataclasses import dataclass
+
+if sys.version_info < (3, 10):
+    from typing import Union
+
+if sys.version_info >= (3, 11):
+    import tomllib
+else:
+    import tomli as tomllib
+
+
+_ContainerT = TypeVar("_ContainerT", bound="ContainerBase")
+_ContainerSectionT = TypeVar("_ContainerSectionT", bound="ContainerSectionBase")
+
+
+_ALL_CONTAINERS: dict[int, Any] = {}
+if sys.version_info >= (3, 10):
+    _ALL_PATHS: dict[int, Path | None] = {}
+else:
+    _ALL_PATHS: dict[int, Union[Path, None]] = {}
+
+
+class FileFormat(str, Enum):
+    "File formats that are supported"
+    TOML = "toml"
+    JSON = "json"
+
+
+@dataclass(frozen=True)
+class ContainerSectionBase:
+    """Base class for all ContainerSection classes (so that we can bound a TypeVar)"""
+
+
+@dataclass(frozen=True)
+class ContainerBase(ABC):
+    """Base class for Config and Settings classes"""
+
+    @classmethod
+    @abstractmethod
+    def kind_string(cls: type[_ContainerT]) -> str:
+        "Return either Config or Settings"
+
+    @classmethod
+    def default_foldername(cls: type[_ContainerT]) -> str:
+        """Return the class name without kind_string, lowercase, with a preceding dot and underscores to seperate words."""
+        kind_str = cls.kind_string()
+        return (
+            "."
+            + sub("(?<!^)(?=[A-Z])", "_", cls.__name__.replace(kind_str, "")).lower()
+        )
+
+    @classmethod
+    def default_filename(cls: type[_ContainerT]) -> str:
+        """Return the kind_string, lowercase, with the extension that fits the file_format."""
+        return f"{cls.kind_string().lower()}.toml"
+
+    if sys.version_info >= (3, 10):
```

Comment:
Whoahh didn't know this worked pretty interesting. One question I have is, is the only difference the `Path | None` vs `Union[ Path, ...]`? I think then I would prefer at the top something like:

```python
if sys.version_info >- (3, 10):
    _PathOrStrT = Path | str
    _PathOptT = Path | None
else:
   _PathOrStrT = Union[Path | str]
   _PathOptT = Union[Path | None]
```

That way we can avoid the multiple implementations I think?

Also should we run mypy for both 3.9 and 3.10/11? Most of the changes seem related to that, and running mypy is quick anyway.

Reply:
If we use this in multiple files it might even make sense to put them in some typing helper file and import them?

Reply:
That is indeed much better, much less code duplication.